src/runtime/string.cpp

```diff
@@ -672,8 +672,70 @@ static PRIMFN(prim_shell_str) {
   RETURN(String::alloc(runtime.heap, shell_escape(str->c_str())));
 }
 
+static PRIMTYPE(type_substr) {
+  TypeVar result;
+  Data::typeResult.clone(result);
+  result[0].unify(Data::typeString);
+  result[1].unify(Data::typeString);
+  return args.size() == 3 &&
+    args[0]->unify(Data::typeString) &&
+    args[1]->unify(Data::typeInteger) &&
+    args[2]->unify(Data::typeInteger) &&
+    out->unify(result);
+}
+
+// Create a 'Result String String' value on the heap.
+Value *allocate_result(Heap &h, bool is_pass, const std::string &msg) {
+    h.reserve(reserve_result() + String::reserve(msg.size()));
+    return claim_result(h, is_pass, String::claim(h, msg));
+}
+
+// This operates on utf8 runes rather than bytes, to match prim_explode.
+static PRIMFN(prim_substr) {
+  EXPECT(3);
+  STRING(input_str, 0);
+  INTEGER_MPZ(pos_arg, 1);
+  INTEGER_MPZ(len_arg, 2);
+
+  int start_pos = mpz_get_si(pos_arg);
+  int length = mpz_get_si(len_arg);
+  if (start_pos < 0 || length < 1) {
+    RETURN(allocate_result(runtime.heap, false, "substr: invalid position or length argument"));
+  }
+
+  std::string new_substr;
+  new_substr.reserve(length);
```

Comment:
So you call `reserve` here but then later on all paths call `reserve` again. You only want to call it once. I'd just reserve the string space at the end with allocate_result instead. The general rule is that any path through a primitive should call reserve at most once.

reserve calls don't stack so you have to actually allocate between reserves or else the reserves are kind of pointless.

Reply:
I think I only call `reserve` once for the string, the other reserve() calls are for the runtime heap

Reply:
Oh I'm a numskull, standard library's string reserve here I see.